Civi/Core/Container.php

```diff
@@ -517,6 +520,8 @@ public static function boot($loadFromDB) {
 
     $bootServices['lockManager'] = self::createLockManager();
 
+    $bootServices['dispatcher.boot']->setDispatchPolicy(\CRM_Core_Config::isUpgradeMode() ? \CRM_Upgrade_DispatchPolicy::get('upgrade.main') : NULL);
+
     if ($loadFromDB && $runtime->dsn) {
```

Comment:
@eileenmcnaughton Yeah, good comment. It is fairly hard to frame a unit-test for this. It took me a bit to think up a QA/defensive mechanism. The primary risk, IMHO, is that some innocuous-looking patch will have some obscure side-effect which inadvertently causes some hook to dispatch in the midst of bootstrap.

(The stuff in a7718acbe04134efd4f7b67626c44d17f26b2ead is a pretty example of otherwise innocuous code which inadvertently dispatches `hook_civicrm_entityTypes`.)

This patch offers a defense using `setDispatchPolicy()`. To see this defense in action, you can sprinkle in calls to `\CRM_Utils_Hook::entityTypes($x);` in various places (e.g. in `createLockManager()` or `SettingsManager::__construct()` or  `CRM_Utils_System_*::initialize()`) - if anything tries to fire a hook before the boot system is good-and-ready for it, then you'll get an exception.

(*Clarification: The `setDispatchPolicy()` notation is probably unfamiliar. The first policy `['/./' => 'not-ready']` matches the wildcard regex and throws a "Not ready exception if any hooks are dispatched. The second policy is `NULL`, which allows all hooks -- notwithstanding the pre-existing edge-case for upgrades.*)

The key thing here is that the defensive measure is *always used* during `boot()`. Consequently, every unit-test and every E2E test will employ this defensive measure. IMHO, the odds are quite high that some automated-test (if not *all automated-tests*) will raise a redflag on regression.

Arguably, the defense should be a little stronger - 42ccedc7c6bdba553c36462d923f887aeb632e18 relaxes the dispatch-policy a little sooner than I'd like. It's just that the branch here (`if ($loadFromDB...)` makes the code a little ugly. I can push up a tweak to tighten it.

Reply:
ab79d6ca29a81d6c23d21f1004bb9b3e4ad92dde extends the defense a little further and expands the comments.